src/node_buffer.cc

```diff
@@ -1051,6 +1054,111 @@ static void EncodeUtf8String(const FunctionCallbackInfo<Value>& args) {
 }
 
 
+static void EncodeInto(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  Isolate* isolate = env->isolate();
+  Local<Context> context = env->context();
+  CHECK_GE(args.Length(), 2);
+  CHECK(args[0]->IsString());
+  CHECK(args[1]->IsUint8Array());
+
+  size_t read = 0;
+  size_t written = 0;
+
+  Utf8Value src(isolate, args[0]);
+  const char* p = *src;
+
+  Local<Uint8Array> dest = args[1].As<Uint8Array>();
+  Local<ArrayBuffer> buf = dest->Buffer();
+  char* write_result =
+      static_cast<char*>(buf->GetContents().Data()) + dest->ByteOffset();
+  size_t dest_length = dest->ByteLength();
+
+  for (size_t i = 0; i < src.length(); ) {
+    char c = p[i];
+    uint32_t code = 0;
+
+    if ((c & 0x80) == 0) {
+      code = c;
+      i++;
+    } else if ((c & 0xE0) == 0xC0) {
+      code = (p[i] & 0x1F) << 6;
+      code |= (p[i+1] & 0x3F);
+      i += 2;
+    } else if ((c & 0xF0) == 0xE0) {
+      code = (p[i] & 0xF) << 12;
+      code |= (p[i+1] & 0x3F) << 6;
+      code |= (p[i+2] & 0x3F);
+      i += 3;
+    } else if ((c & 0xF8) == 0xF0) {
+      code = (p[i] & 0x7) << 18;
+      code |= (p[i+1] & 0x3F) << 12;
+      code |= (p[i+2] & 0x3F) << 6;
+      code |= (p[i+3] & 0x3F);
+      i += 4;
+    } else if ((c & 0xFC) == 0xF8) {
+      code = (p[i] & 0x3) << 24;
+      code |= (p[i] & 0x3F) << 18;
+      code |= (p[i] & 0x3F) << 12;
+      code |= (p[i] & 0x3F) << 6;
+      code |= (p[i] & 0x3F);
+      i += 5;
+    } else if ((c & 0xFE) == 0xFC) {
+      code = (p[i] & 0x1) << 30;
+      code |= (p[i] & 0x3F) << 24;
+      code |= (p[i] & 0x3F) << 18;
+      code |= (p[i] & 0x3F) << 12;
+      code |= (p[i] & 0x3F) << 6;
+      code |= (p[i] & 0x3F);
+      i += 6;
+    }
+
+    if (code <= 0x7F) {
+      if (dest_length < 1) break;
+      read += 1;
+      written += 1;
+      dest_length -= 1;
+    } else if (code <= 0x7FF) {
+      if (dest_length < 2) break;
+      read += 1;
+      written += 2;
+      dest_length -= 2;
+    } else if (code <= 0xFFFF) {
+      if (dest_length < 3) break;
+      read += 1;
+      written += 3;
+      dest_length -= 3;
+    } else if (code <= 0x1FFFFF) {
+      if (dest_length < 4) break;
+      read += 2;
+      written += 4;
+      dest_length -= 4;
+    } else {
+      // invalid unicode
+    }
+  }
+
+  Local<String> str = args[0].As<String>();
+  str->WriteUtf8(isolate,
+                 write_result,
+                 written,
+                 nullptr,
+                 String::NO_NULL_TERMINATION | String::REPLACE_INVALID_UTF8);
```

Comment:
We’re still UTF-8-encoding the string twice, once here and once with `Utf8Value` (and then re-decoding the latter value again) … can we avoid that? We can look at the individual code units from JS, too, so I think we could compute `read` and `written` there?